### PR TITLE
coordinated forks update

### PR DESCRIPTION
**Pull request by Theoklitos**
Added two cli commands:
1. Vote
2. Cancel your vote

[They are part of ValidatorConfig.py](https://github.com/radixdlt/node-runner/blob/37280e8098b0c7c119375e015aff197fe6f0e88e/node-runner-cli/api/ValidatorConfig.py#L118). Run them with:
```
python radixnode.py api core update-validator-config
```
The CLI text when a vote is required:
vote_required

After a vote has been cast:
no_vote_required

When there is no candidate branch:
nothing_needed

Users are warned if they register:
warning

And if they want to update metadata:
warning2

Voting (and coordinated forks in general) is explained here: [Design Doc](https://radixdlt.atlassian.net/wiki/spaces/RPNV1/pages/2606071814/Coordinated+forks+design#How-does-a-validator-send-their-vote?).

I have tested the above in a local network running the `feature/NT-242-NT-247-NT-248-fork-voting` branch.

### CODE REVIEW

File: node-runner-cli/Pipfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-radixdlt_core_client = {git = "https://github.com/radixdlt/python-core-client", editable = true, ref = "1.0-beta.1"}
+core_client = {git = "https://github.com/radixdlt/python-core-client", editable = true}
```
Comment (dhedey) [inline node-runner-cli/Pipfile:7]: Can we specify an _exact_ tag ref for this? 👍  - to avoid other people or the build pipeline building with a cached / incorrect version?

Comment (shambupujar) [inline node-runner-cli/Pipfile:7]: you may need to add the reference back , otherwise, version changes will break the cli

**Review by shambupujar — APPROVED** [review]

File: node-runner-cli/api/ValidatorConfig.py [hunk 2]
```diff
@@ -108,12 +108,30 @@ def add_change_ownerid(actions: List, validator_info: EntityResponse):
...
                 return actions
             Helpers.print_coloured_line("Owner entered is same . So action will not be applied", bcolors.WARNING)
 
         return actions
 
+    @staticmethod
+    def vote(actions: List, validator_info: EntityResponse, health, engine_configuration):
+        print("-------- Vote -----\n")
+        latest_fork_name = print_vote_and_fork_info(health, engine_configuration)        
+        response = input(f"{bcolors.BOLD}\nDo you want to cast a vote for fork '{latest_fork_name}' [y/n]? " +
```
Comment (dhedey) [inline node-runner-cli/api/ValidatorConfig.py:121]: The more I think about this, the more I'm sure that we should just separate vote casting from registration. Trying to explain to the user here that by voting, they're also registering, feels quite weird / hard to justify?

Why not just have them as separate commands at construction? And then the CLI can make the choice if it wishes to combine them for the user?

What are your thoughts @LukasGasior1  and @Theoklitos ?

File: node-runner-cli/api/ValidatorConfig.py [hunk 3]
```diff
@@ -108,12 +108,30 @@ def add_change_ownerid(actions: List, validator_info: EntityResponse):
...
-                actions.append(Action().set_validator_owner(owner_id))
+                actions.append(Action.set_validator_owner(owner_id))
                 return actions
             Helpers.print_coloured_line("Owner entered is same . So action will not be applied", bcolors.WARNING)
 
         return actions
 
+    @staticmethod
+    def vote(actions: List, validator_info: EntityResponse, health, engine_configuration):
+        print("-------- Vote -----\n")
```
Comment (dhedey) [inline node-runner-cli/api/ValidatorConfig.py:119]: The neat part of my head is annoyed that the number of dashes on the left and right of this are uneven :')

File: node-runner-cli/api/ValidatorConfig.py [hunk 4]
```diff
@@ -108,12 +108,30 @@ def add_change_ownerid(actions: List, validator_info: EntityResponse):
...
+        response = input(f"{bcolors.BOLD}\nDo you want to cast a vote for fork '{latest_fork_name}' [y/n]? " +
+                         f"Presssing Enter does not cast a vote: {bcolors.ENDC}").strip()
+        if response.lower() == 'y':
+            actions.append(Action.vote())
+        return actions
+    
+    @staticmethod
+    def cancel_vote(actions: List, validator_info: EntityResponse):
+        response = input(f"{bcolors.BOLD}\nDo you want to cancel your vote [y/n]? " +
+                         f"Presssing Enter does not cancel your vote: {bcolors.ENDC}").strip()
```
Comment (dhedey) [inline node-runner-cli/api/ValidatorConfig.py:130]: `Pressing`*

File: node-runner-cli/api/ValidatorConfig.py [hunk 5]
```diff
@@ -108,12 +108,30 @@ def add_change_ownerid(actions: List, validator_info: EntityResponse):
...
             Helpers.print_coloured_line("Owner entered is same . So action will not be applied", bcolors.WARNING)
 
         return actions
 
+    @staticmethod
+    def vote(actions: List, validator_info: EntityResponse, health, engine_configuration):
+        print("-------- Vote -----\n")
+        latest_fork_name = print_vote_and_fork_info(health, engine_configuration)        
+        response = input(f"{bcolors.BOLD}\nDo you want to cast a vote for fork '{latest_fork_name}' [y/n]? " +
+                         f"Presssing Enter does not cast a vote: {bcolors.ENDC}").strip()
```
Comment (dhedey) [inline node-runner-cli/api/ValidatorConfig.py:122]: Pressing*. Also, what do you mean pressing enter does not cast a vote? Do you mean that entering "" counts as "n"?

File: node-runner-cli/utils/utils.py [hunk 6]
```diff
@@ -33,6 +33,25 @@ def run_shell_command(cmd, env=None, shell=False, fail_on_error=True, quite=Fals
         sys.exit()
     return result
 
+def print_vote_and_fork_info(health, engine_configuration):
+    vote_status = health['fork_vote_status']
+    print(f"Vote status : {vote_status}")
```
Comment (dhedey) [inline node-runner-cli/utils/utils.py:38]: The spaces before the colons in these messages look a little weird to me

File: node-runner-cli/utils/utils.py [hunk 7]
```diff
@@ -33,6 +33,25 @@ def run_shell_command(cmd, env=None, shell=False, fail_on_error=True, quite=Fals
...
+    if vote_status == 'VOTE_REQUIRED':
+        print(f"{bcolors.WARNING}Your vote is required{bcolors.ENDC}")    
+    else:
+        print(f"{bcolors.WARNING}No vote is required at the moment{bcolors.ENDC}")    
+    latest_fork = engine_configuration['forks'][-1]
```
Comment (dhedey) [inline node-runner-cli/utils/utils.py:43]: We're using the term "Newest Fork" for this now in the codebase

File: node-runner-cli/utils/utils.py [hunk 8]
```diff
@@ -33,6 +33,25 @@ def run_shell_command(cmd, env=None, shell=False, fail_on_error=True, quite=Fals
...
+    else:
+        print(f"{bcolors.WARNING}No vote is required at the moment{bcolors.ENDC}")    
+    latest_fork = engine_configuration['forks'][-1] 
+    latest_fork_name = latest_fork['name']
+    is_candidate = latest_fork['is_candidate']
+    if health['current_fork_name'] == latest_fork_name:        
+        print(f"\nCurrent fork : {latest_fork_name}, is candidate: {is_candidate}")    
+        print(f"{bcolors.WARNING}The validator is on the latest fork{bcolors.ENDC}")
+    else:
+        print(f"\nLatest fork : {latest_fork_name}, is candidate: {is_candidate}")
```
Comment (dhedey) [inline node-runner-cli/utils/utils.py:50]: Can we add a print line explaining "The network is currently running fork X"

File: node-runner-cli/utils/utils.py [hunk 9]
```diff
@@ -33,6 +33,25 @@ def run_shell_command(cmd, env=None, shell=False, fail_on_error=True, quite=Fals
...
+    if vote_status == 'VOTE_REQUIRED':
+        print(f"{bcolors.WARNING}Your vote is required{bcolors.ENDC}")    
+    else:
+        print(f"{bcolors.WARNING}No vote is required at the moment{bcolors.ENDC}")    
+    latest_fork = engine_configuration['forks'][-1] 
+    latest_fork_name = latest_fork['name']
+    is_candidate = latest_fork['is_candidate']
+    if health['current_fork_name'] == latest_fork_name:        
+        print(f"\nCurrent fork : {latest_fork_name}, is candidate: {is_candidate}")    
+        print(f"{bcolors.WARNING}The validator is on the latest fork{bcolors.ENDC}")
```
Comment (dhedey) [inline node-runner-cli/utils/utils.py:48]: This is probably more accurately "The validator is currently running its newest fork"

File: node-runner-cli/utils/utils.py (hunk 8 above)
Comment (dhedey) [reply node-runner-cli/utils/utils.py:50]: Instead of is_candidate, can we display either "this is a candidate fork, which needs X% vote to enact" or "this is a fixed epoch fork, which will be enacted at epoch X" (if we have this detail?)

File: node-runner-cli/api/ValidatorConfig.py (hunk 2 above)
Comment (dhedey) [reply node-runner-cli/api/ValidatorConfig.py:121]: (Just to check - does Action.vote currently register the validator at the moment? Does registering or updating metadata cause a voting action?)

**Review by dhedey — COMMENTED** [review]
Thanks Theo! :) 👍

File: node-runner-cli/Pipfile (hunk 1 above)
Comment (Theoklitos) [reply node-runner-cli/Pipfile:7]: right

File: node-runner-cli/api/ValidatorConfig.py (hunk 2 above)
Comment (Theoklitos) [reply node-runner-cli/api/ValidatorConfig.py:121]: I agree that we should separate vote casting from registration, I will push a change for this

>  does Action.vote currently register the validator at the moment? Does registering or updating metadata cause a voting action?

Yes Action.vote registers the validator. Register _and_ updating metadata cause a voting action. But there is also a third way: Submitting an empty ValidatorSystemMetadata. [Design doc link](https://radixdlt.atlassian.net/wiki/spaces/RPNV1/pages/2606071814/Coordinated+forks+design#How-does-a-validator-change-%2F-remove-their-vote%3F)

File: node-runner-cli/Pipfile (hunk 1 above)
Comment (Theoklitos) [reply node-runner-cli/Pipfile]: 👍

File: node-runner-cli/utils/utils.py (hunk 8 above)
Comment (Theoklitos) [reply node-runner-cli/utils/utils.py]: Unfortunately we do not have this information. Is it worth including it e.g. under /engine/configuration?

File: node-runner-cli/Pipfile [hunk 10]
```diff
@@ -4,7 +4,7 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-radixdlt_core_client = {git = "https://github.com/radixdlt/python-core-client", editable = true, ref = "1.0-beta.1"}
+core_client = {git = "https://github.com/radixdlt/python-core-client", editable = true, ref = "master"}
```
Comment (dhedey) [inline node-runner-cli/Pipfile:7]: We should not use `master` or a branch here as it's not immutable and it breaks mutability assumptions in the package resolution, and for people downloading the package and expecting it to work.

Please create a tag for a versioned release in that repo and refer to the branch here :)

File: node-runner-cli/radixnode.py [hunk 11]
```diff
@@ -24,7 +24,7 @@
 from setup import Base, Docker, SystemD
 from utils.utils import Helpers
 from utils.utils import run_shell_command
-from version import __version__
+#from version import __version__
```
Comment (dhedey) [inline node-runner-cli/radixnode.py:27]: This commented line should be removed.

Out of interest, why was this removed? Or was it giving us the python version, not the package verison? :)

File: node-runner-cli/utils/utils.py [hunk 12]
```diff
@@ -33,25 +33,29 @@ def run_shell_command(cmd, env=None, shell=False, fail_on_error=True, quite=Fals
         sys.exit()
     return result
 
+def check_for_candidate_forks(health):
```
Comment (dhedey) [inline node-runner-cli/utils/utils.py:36]: Nice, thanks 👍

**Review by dhedey — CHANGES_REQUESTED** [review]
Looks good - a couple of minor tweaks left

File: node-runner-cli/radixnode.py (hunk 11 above)
Comment (Theoklitos) [reply node-runner-cli/radixnode.py:27]: I removed it by accident...

**Review by dhedey — APPROVED** [review]
